ui/src/lib/yjs/useYWorkflow.ts

```diff
@@ -135,9 +135,51 @@ export default ({
       handleWorkflowIdChange,
     ],
   );
+
+  const handleWorkflowRename = useCallback(
+    (id: string, name: string) =>
+      undoTrackerActionWrapper(() => {
+        if (!name.trim()) {
+          throw new Error("Workflow name cannot be empty");
+        }
+
+        // Update local state
+        setWorkflows((w) => w.map((w) => (w.id === id ? { ...w, name } : w)));
+
+        const workflowIndex = workflows.findIndex((w) => w.id === id);
+
+        // Update Yjs shared data
+        const workflow = yWorkflows.get(workflowIndex);
+        if (!workflow) {
+          throw new Error("Workflow not found");
+        }
+        workflow.set("name", new YText(name));
+
+        // Update subworkflow node in main workflow if this is a subworkflow
+        if (workflowIndex > 0) {
+          const mainWorkflow = yWorkflows.get(0);
+          const mainWorkflowNodes = mainWorkflow?.get("nodes") as YNodesArray;
+          const workflowId = workflows[workflowIndex].id;
+
+          mainWorkflowNodes?.forEach((node, index) => {
+            if (node.id === workflowId) {
+              const updatedNode = {
+                ...node,
+                data: { ...node.data, name },
+              };
+              mainWorkflowNodes.delete(index);
+              mainWorkflowNodes.insert(index, [updatedNode]);
+            }
+          });
+        }
```

Comment:
I'm not sure we really need to do any index checking (anyways, the local state is already updated so putting the check here would only potentially mess up synchronization with local state and the yDoc).
Also, any reason at this point you can't use handleNodesUpdate? Might be able to simplify code here a bit and reduce similar code

Reply:
I removed the index checking. I tried using the handleNodesUpdate but it didn't work probably because the undoActionsHandler loops inside.

Reply:
Just removing the `undoTrackerActionWrapper` that is wrapping `hadnleWorkflowRename` and then using `handleNodesUpdate` should work (in my mind). 
There is a undo tracker provider attached to the y.doc, and then we use `undoTrackerActionWrapper` to watch specific parts of the y.doc we want to listen to (and the rest should be ignored).Since the rest of handleWorkflowRename has nothing to do with Yjs, this should work as expected

Reply:
As discussed this creates a circular dependency issue